language/diem-framework/modules/0L/RecoveryMode.move

```diff
@@ -0,0 +1,90 @@
+///////////////////////////////////////////////////////////////////////////
+// 0L Module
+// Recovery Mode
+///////////////////////////////////////////////////////////////////////////
+// For when an admin upgrade or network halt recovery needs to be exectuted.
+// For use for example in preventing front running by miners and validators 
+// for rewards while the network is unstable.
+///////////////////////////////////////////////////////////////////////////
+
+
+
+address 0x1 {
+module RecoveryMode {
+
+    use 0x1::CoreAddresses;
+    use 0x1::DiemConfig;
+    use 0x1::DiemSystem;
+    use 0x1::Vector;
+    use 0x1::Testnet;
+    use 0x1::StagingNet;
+
+    struct RecoveryMode has copy, key, drop, store{
+      // set this if a validator set needs to be overriden
+      // if list is empty, it will use validator set.
+      fixed_set: vector<address>, 
+      epoch_ends: u64,
+    }
+
+    // private function so that it can only be called by vm session.
+    // should never be used in production.
+    fun init_recovery(vm: &signer, vals: vector<address>, epoch_ends: u64) {
+      if (!is_recovery()) {
+        move_to<RecoveryMode>(vm, RecoveryMode {
+          fixed_set: vals,
+          epoch_ends,
+        });
+      }
+    }
+
+    public fun maybe_remove_debug_at_epoch(vm: &signer) acquires RecoveryMode {
+      CoreAddresses::assert_vm(vm);
+      if (!exists<RecoveryMode>(CoreAddresses::VM_RESERVED_ADDRESS())) return;
+
+      let enough_vals = if (
+        Testnet::is_testnet() || 
+        StagingNet::is_staging_net()
+      ){ true }
+      else { (DiemSystem::validator_set_size() >= 21) };
```

Comment:
Does this function get the current validator set size, or the number of eligible validators? if it's the former, then enough_vals will never be true unless the original fixed set is over 21, as I don't see a mechanism to add validators to the fixed set.

Reply:
Good point.

My assumption is that there are 2 modes, and they are exclusive:
1. where you create a fixed set for debugging.
2. you are actually going to recover so the fixed set list is empty.

Originally I wrote #1 because it wasn't clear how we could jailbreak a stuck netowork. TL;DR we may just do without the fixed_set implementation. And REcoveryMode then only does the grace period before payments start again.

Reply:
We could just have the fixed_case expire after # of epochs. So we keep this option for future cases

Reply:
see now

Reply:
Yes, this looks better